File: crates/fno-agents/src/spawn_payload.rs
pub const BREVITY_MARKER: &str = "<fno_relay_compression>";
pub const BREVITY_END_MARKER: &str = "</fno_relay_compression>";
pub const BREVITY_INSTRUCTION: &str = "Keep reports and handoffs at 80 words or fewer unless this payload requires a longer artifact or exact output schema. Think fully; write only the requested result, essential reason, and next action. Drop filler, pleasantries, hedges, repeated context, and articles where clear. Fragments work. Keep technical terms, commands, errors, numbers, units, negation, and code blocks exact. Put long detail in durable artifacts when available; return a path or link.";

pub fn enrich_spawn_payload(message: &str) -> String {
    let block = format!("{BREVITY_MARKER}\n{BREVITY_INSTRUCTION}\n{BREVITY_END_MARKER}");
    if message.is_empty() || message.contains(&block) {
        return message.to_owned();
    }
    format!("{message}\n\n{block}")
}

#[cfg(test)]
mod tests {
    use super::{enrich_spawn_payload, BREVITY_END_MARKER, BREVITY_INSTRUCTION, BREVITY_MARKER};

    #[test]
    fn nonempty_payload_keeps_original_and_appends_positive_marker_once() {
        let original = "Run `cargo test`, not `cargo check`. Keep 80, not 81.";
        let enriched = enrich_spawn_payload(original);

        assert!(enriched.starts_with(&format!("{original}\n\n")));
        assert_eq!(enriched.matches(BREVITY_MARKER).count(), 1);
    }

    #[test]
    fn empty_and_already_enriched_payloads_are_unchanged() {
        assert_eq!(enrich_spawn_payload(""), "");
        let enriched = enrich_spawn_payload("report");
        assert_eq!(enrich_spawn_payload(&enriched), enriched);
    }

    #[test]
    fn incidental_marker_text_still_gets_the_complete_guidance_block() {
        let original = format!("Explain the literal {BREVITY_MARKER} token.");
        let enriched = enrich_spawn_payload(&original);
        assert!(enriched.starts_with(&format!("{original}\n\n")));
        assert_eq!(enriched.matches(BREVITY_MARKER).count(), 2);
        assert!(enriched.ends_with(BREVITY_END_MARKER));
    }

    #[test]
    fn runtime_constants_match_the_shared_cross_language_fixture() {
        const FIXTURE: &str = include_str!(concat!(
            env!("CARGO_MANIFEST_DIR"),
            "/../../schemas/spawn-brevity.json"
        ));
        let fixture: serde_json::Value = serde_json::from_str(FIXTURE).expect("valid fixture");

        assert_eq!(fixture["marker"], BREVITY_MARKER);
        assert_eq!(fixture["end_marker"], BREVITY_END_MARKER);
        assert_eq!(fixture["instruction"], BREVITY_INSTRUCTION);
    }
}

File: cli/src/fno/agents/spawn_payload.py
"""Portable output-compression guidance for spawned workers."""

from __future__ import annotations

import json
import os
from pathlib import Path
from typing import Optional


BREVITY_MARKER = "<fno_relay_compression>"
BREVITY_END_MARKER = "</fno_relay_compression>"
BREVITY_INSTRUCTION = (
    "Keep reports and handoffs at 80 words or fewer unless this payload requires a "
    "longer artifact or exact output schema. Think fully; write only the requested "
    "result, essential reason, and next action. Drop filler, pleasantries, hedges, "
    "repeated context, and articles where clear. Fragments work. Keep technical terms, "
    "commands, errors, numbers, units, negation, and code blocks exact. Put long detail "
    "in durable artifacts when available; return a path or link."
)
BREVITY_BLOCK = f"{BREVITY_MARKER}\n{BREVITY_INSTRUCTION}\n{BREVITY_END_MARKER}"

# The spawner arms the current attempt's bound binding here (absolute path to
# the bound JSON under the existing artifact root). Both launch substrates read
# the same variable through prepare_spawn_payload, so pane and non-pane carries
# stay identical by construction.
TASK_CONTEXT_ENV = "FNO_TASK_CONTEXT_FILE"

# Bounded by declaration: at most this many declared constraints ride a
# payload. Source CONTENTS never ride a payload at all.
MAX_PAYLOAD_CONSTRAINTS = 10

TASK_CONTEXT_TAG = "<task-context "


def enrich_spawn_payload(message: str) -> str:
    """Append first-party brevity guidance once to a non-empty spawn payload."""
    if not message or BREVITY_BLOCK in message:
        return message
    return f"{message}\n\n{BREVITY_BLOCK}"


def load_task_context(env: Optional[str] = None) -> Optional[dict]:
    """The spawner's bound binding, when FNO_TASK_CONTEXT_FILE names a readable one.

    A missing, unreadable, or malformed file loads as UNSET (no block rides the
    payload); it never renders as a binding. A declared-but-corrupt binding is
    the native verifier's refusal to report, not this loader's guess.
    """
    path = (env if env is not None else os.environ.get(TASK_CONTEXT_ENV)) or ""
    if not path.strip():
        return None
    try:
        raw = json.loads(Path(path).read_text(encoding="utf-8"))
    except (OSError, ValueError):
        return None
    return raw if isinstance(raw, dict) else None


def task_context_block(binding: dict) -> Optional[str]:
    """The bounded pointer block: identity + digest + declared constraints.

    Never source contents, never a second skill preamble. The pointer is not a
    read: having this block in a payload proves nothing about what the
    recipient has read (the stage field stays the honest record).
    """
    node = binding.get("node")
    attempt = binding.get("attempt")
    digest = binding.get("binding_digest")
    if not (isinstance(node, str) and node and isinstance(attempt, str) and attempt):
        return None
    if not (isinstance(digest, str) and digest):
        return None
    stage = binding.get("stage") or "prepared"
    constraints = [
        c
        for c in binding.get("required_constraints", [])
        if isinstance(c, str) and c.strip()
    ][:MAX_PAYLOAD_CONSTRAINTS]
    lines = [
        f'{TASK_CONTEXT_TAG}node="{node}" attempt="{attempt}" '
        f'binding_digest="{digest}" stage="{stage}">'
    ]
    if constraints:
        lines.append("Required constraints:")
        lines.extend(f"- {c}" for c in constraints)
    lines.append("</task-context>")
    return "\n".join(lines)


def prepare_spawn_payload(
    message: str, task_context: Optional[dict] = None
) -> tuple[str, dict]:
    """ONE payload-preparation entry for both launch substrates (x-59b0).

    Preserves the original (normalized) message as the prefix, appends the
    brevity guidance once, then the bounded task-context pointer once. Returns
    (payload, measures); measures keeps the two budgets separate -
    payload_bytes is what THIS submission carries, not the sources' bytes.
    """
    payload = enrich_spawn_payload(message)
    binding = task_context if task_context is not None else load_task_context()
    measures: dict = {"task_context": False, "payload_bytes": len(payload.encode("utf-8"))}
    if binding:
        block = task_context_block(binding)
        if block and TASK_CONTEXT_TAG not in payload:
            payload = f"{payload}\n\n{block}"
            measures["task_context"] = True
        measures["payload_bytes"] = len(payload.encode("utf-8"))
    return payload, measures
